### PR TITLE
2.0.0 typechecker 88

### PR DESCRIPTION
**Pull request by CThuleHansen**
# Typechecker version alpha 1.0.0. 

It has been tested with a fully expanded specification originating from `initialize_fixedstep_unfold` test in `fullspectest`. Available in `tests` of the `typechecker` project.

## Syntax Changes 
The array syntax is: `int a[4]` for variable declarations and `int a[]` for formal parameters. `int[] a[4]` and `int[] a` and `int[] a[]` does **NOT** parse anymore.

## Parts not considered (Fix on Friday 13th?)
- How shall import modules be added to the typechecker?
- Expansion plugins?
- Error codes
- Map between nodes in AST and their types, such that one can look it up
- Tests in Maestro module are **NOT** typechecked
- Typechecker is **NOT** enabled in general

### CODE REVIEW

File: ast/src/main/resources/mabl.astv2 [hunk 1]
```diff
@@ -68,7 +68,8 @@ parameter
 
 
 
-exp
+exp { ->
+    | (type):type }
```
[the remark this answers is not in the document]
Comment (CThuleHansen) [reply ast/src/main/resources/mabl.astv2:72]: Removed

File: ast/src/main/resources/mabl.astv2.tostring [hunk 2]
```diff
@@ -14,6 +14,8 @@ import java.util.stream.Collectors;
 
 %declaration->function = [returnType] +$ " "+ $[name]$+"("+ $[formals]$.stream().map(Object::toString).collect(Collectors.joining(", "))$+")"
 %declaration->variable = [type]+ $ " "+ $[name]$ + ""+($[size]$==null || $[isArray]$==null || !$[isArray]$?"":$[size]$.stream().map(s->s.toString()).collect(Collectors.joining(",","[","]")))+ ( $[initializer]$ ==null?"":" = " +$[initializer]$)$
+//%declaration->variable = "" + $ org.intocps.maestro.ast.display.PrintVariableDeclaration.VariableDeclarationToString(this) + "" + ( $[initializer]$ ==null?"":" = " +$[initializer]$) $
```
[the remark this answers is not in the document]
Comment (CThuleHansen) [reply ast/src/main/resources/mabl.astv2.tostring:17]: Removed

File: maestro/src/main/java/org/intocps/maestro/TypecheckerTest.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+     *
+     * @param node
+     * @param question
+     * @return
+     * @throws AnalysisException
+     */
+    public PType caseBinaryExpBase(SBinaryExpBase node, TypeCheckInfo question) throws AnalysisException {
+        PType leftType = node.getLeft().apply(this, question);
+        PType rightType = node.getLeft().apply(this, question);
+        if (leftType.getClass().equals(rightType.getClass()) && leftType instanceof SNumericPrimitiveBase) {
```
[the remark this answers is not in the document]
Comment (CThuleHansen) [reply maestro/src/main/java/org/intocps/maestro/TypecheckerTest.java:67]: Was never used

File: maestro/src/main/java/org/intocps/maestro/TypecheckerTest.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+            node.setType(type);
+            return type;
+        } else {
+            throw new AnalysisException("typechecker: caseBinaryExpBase");
+        }
+    }
+
+    @Override
+    public PType caseALessBinaryExp(ALessBinaryExp node, TypeCheckInfo question) throws AnalysisException {
+        return caseBinaryExpBase(node, question);
```
[the remark this answers is not in the document]
Comment (CThuleHansen) [reply maestro/src/main/java/org/intocps/maestro/TypecheckerTest.java:78]: Class was never used

File: typechecker/src/main/java/org/intocps/maestro/typechecker/TypeComparator.java [hunk 5]
```diff
@@ -10,12 +10,17 @@
 public class TypeComparator {
```
[the remark this answers is not in the document]
Comment (CThuleHansen) [reply typechecker/src/main/java/org/intocps/maestro/typechecker/TypeComparator.java:10]: TypeCheckerTest is removed

File: typechecker/src/test/java/ErrorReporter.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+import org.intocps.maestro.core.messages.MableError;
+import org.intocps.maestro.core.messages.MableMessage;
+import org.intocps.maestro.core.messages.MableWarning;
+import org.intocps.maestro.typechecker.InternalException;
+
+import java.io.PrintWriter;
+import java.util.List;
+import java.util.Vector;
+
+public class ErrorReporter implements IErrorReporter {
```
[the remark this answers is not in the document]
Comment (CThuleHansen) [reply typechecker/src/test/java/ErrorReporter.java:14]: Duplicate. Where should we place it?

File: typechecker/src/test/java/parseFmi2Module.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.Arrays;
+import java.util.List;
+
+public class parseFmi2Module {
+
+    @Test
```
[the remark this answers is not in the document]
Comment (CThuleHansen) [reply typechecker/src/test/java/parseFmi2Module.java:17]: Agreed

File: typechecker/src/test/java/ErrorReporter.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (CThuleHansen) [reply typechecker/src/test/java/ErrorReporter.java:14]: Fixed

File: ast/src/main/resources/mabl.astv2 (hunk 1 above)
[the remark this answers is not in the document]
Comment (nickbattle) [reply ast/src/main/resources/mabl.astv2:72]: Yes! This is what VDMJ cleaned up in version 4. The AST is immutable; TC trees (and others) contain extra information created by different analyses.

**Pull request merged**